File: speedviewlib/src/main/java/com/github/anastr/speedviewlib/components/note/TextNote.java
package com.github.anastr.speedviewlib.components.note;

import android.content.Context;
import android.graphics.Canvas;
import android.graphics.Paint;
import android.graphics.Typeface;
import android.text.Layout;
import android.text.StaticLayout;
import android.text.TextPaint;

/**
 * this Library build By Anas Altair
 * see it on <a href="https://github.com/anastr/SpeedView">GitHub</a>
 */
public class TextNote extends Note<TextNote> {

    private CharSequence noteText;
    private TextPaint notePaint = new TextPaint(Paint.ANTI_ALIAS_FLAG);
    private float textSize = notePaint.getTextSize();
    private StaticLayout textLayout;

    /**
     * @param context you can use {@code getApplicationContext()} method.
     * @param noteText text to display, support SpannableString and multi-lines.
     */
    public TextNote(Context context, CharSequence noteText) {
        super(context);
        if (noteText == null)
            throw new IllegalArgumentException("noteText cannot be null.");
        this.noteText = noteText;
        notePaint.setTextAlign(Paint.Align.LEFT);
    }

    @Override
    public void build(int viewWidth) {
        textLayout = new StaticLayout(noteText, notePaint, viewWidth
                , Layout.Alignment.ALIGN_NORMAL, 1.0f, 0.0f, false);
        int w = 0;
        for (int i = 0; i< textLayout.getLineCount(); i++)
            w = (int) Math.max(w, textLayout.getLineWidth(i));
        noticeContainsSizeChange(w, textLayout.getHeight());
    }

    @Override
    protected void drawContains(Canvas canvas, float leftX, float topY) {
        canvas.save();
        canvas.translate(leftX, topY);
        textLayout.draw(canvas);
        canvas.restore();
    }

    public float getTextSize() {
        return textSize;
    }

    /**
     * set Text size.
     * @param textSize in Pixel.
     * @return This Note object to allow for chaining of calls to set methods.
     */
    public TextNote setTextSize(float textSize) {
        this.textSize = textSize;
        notePaint.setTextSize(textSize);
        return this;
    }

    /**
     * to change font or text style.
     * @param typeface new Typeface.
     * @return This Note object to allow for chaining of calls to set methods.
     */
    public TextNote setTextTypeFace(Typeface typeface) {
        notePaint.setTypeface(typeface);
        return this;
    }

    public int getTextColor() {
        return notePaint.getColor();
    }

    /**
     * set text color.
     * @param textColor new color.
     * @return This Note object to allow for chaining of calls to set methods.
     */
    public TextNote setTextColor(int textColor) {
        notePaint.setColor(textColor);
        return this;
    }
}

File: speedviewlib/src/main/java/com/github/anastr/speedviewlib/components/note/TextNote.kt
package com.github.anastr.speedviewlib.components.note

import android.content.Context
import android.graphics.Canvas
import android.graphics.Paint
import android.graphics.Typeface
import android.text.Layout
import android.text.StaticLayout
import android.text.TextPaint
import kotlin.math.max

/**
 * this Library build By Anas Altair
 * see it on [GitHub](https://github.com/anastr/SpeedView)
 */
class TextNote
/**
 * @param context you can use `getApplicationContext()` method.
 * @param noteText text to display, support SpannableString and multi-lines.
 */
(context: Context, private val noteText: CharSequence?) : Note<TextNote>(context) {
    private val notePaint = TextPaint(Paint.ANTI_ALIAS_FLAG)
    private var textSize = notePaint.textSize
    private var textLayout: StaticLayout? = null

    val textColor: Int
        get() = notePaint.color

    init {
        requireNotNull(noteText) { "noteText cannot be null." }
        notePaint.textAlign = Paint.Align.LEFT
    }

    override fun build(viewWidth: Int) {
        textLayout = StaticLayout(noteText, notePaint, viewWidth, Layout.Alignment.ALIGN_NORMAL, 1.0f, 0.0f, true)
        var w = 0
        for (i in 0 until textLayout!!.lineCount)
            w = max(w.toFloat(), textLayout!!.getLineWidth(i)).toInt()
        noticeContainsSizeChange(w, textLayout!!.height)
    }

    override fun drawContains(canvas: Canvas, leftX: Float, topY: Float) {
        canvas.save()
        canvas.translate(leftX, topY)
        textLayout!!.draw(canvas)
        canvas.restore()
    }

    fun getTextSize(): Float {
        return textSize
    }

    /**
     * set Text size.
     * @param textSize in Pixel.
     * @return This Note object to allow for chaining of calls to set methods.
     */
    fun setTextSize(textSize: Float): TextNote {
        this.textSize = textSize
        notePaint.textSize = textSize
        return this
    }

    /**
     * to change font or text style.
     * @param typeface new Typeface.
     * @return This Note object to allow for chaining of calls to set methods.
     */
    fun setTextTypeFace(typeface: Typeface): TextNote {
        notePaint.typeface = typeface
        return this
    }

    /**
     * set text color.
     * @param textColor new color.
     * @return This Note object to allow for chaining of calls to set methods.
     */
    fun setTextColor(textColor: Int): TextNote {
        notePaint.color = textColor
        return this
    }
}
